### PR TITLE
Clean Up Repository and Add Types

### PR DESCRIPTION
**Pull request by squat**
- meowlflow: remove unused module
  Remove the loghandler module as it is not used in this project.
- dependencies: add mypy
  Add mypy for type checking.
- *: add types
  Finish adding type annotations to the entire project.
- README: regenerate
- Makefile: simplify Makefile

### CODE REVIEW

File: examples/document_splitter_schema.py [hunk 1]
```diff
@@ -18,7 +19,7 @@
 class Request(BaseRequest):
     __root__: List[Text]
 
-    def transform(self):
+    def transform(self) -> Any:
```
Comment (dswah) [inline examples/document_splitter_schema.py:22]: why are we doing json dumps here?
I recall that we removed this in our [recent schemas](https://github.com/mietright/conny-ml-classification/blob/main/conny_ml/classification/schemas/giro_document_splitting.py)
because we simplified the corresponding code in the schema handler: https://github.com/connylabs/meowlflow/blob/204d1255d01fef4e6c880ed646b1d0938a1b1ebf/meowlflow/sidecar.py#L172

File: meowlflow/promote.py [hunk 2]
```diff
@@ -90,7 +93,7 @@ def register_model(run, model_name):
 @click.option(
     "--direction",
     default="maximize",
-    type=click.Choice(_COMPARE.keys(), case_sensitive=False),
+    type=click.Choice(list(_COMPARE.keys()), case_sensitive=False),
```
Comment (dswah) [inline meowlflow/promote.py:96]: :+1:

File: meowlflow/promote.py [hunk 3]
```diff
@@ -1,6 +1,10 @@
 import sys
+from typing import Optional
 
 import click
+from mlflow.entities import Run
+from mlflow.tracking._model_registry.client import ModelRegistryClient
+from mlflow.entities.model_registry import ModelVersion
```
Comment (dswah) [inline meowlflow/promote.py:7]: wow good find!

**Review by dswah — APPROVED** [review]
looks great lucas!
nice work!
this looks very messy to clean up!

a couple of questions:

1. what to do with the `request.transform` and the `json.dumps` ? I recall that we had opted to remove that bit since we could achieve the desired behavior without encoding/decoding (which is less confusing too)
2. how do you update the `schema.json` files  after making a change to one of the `Schema.py` files? is this automated? If not, what is the process? (and is the process documented in the readme?

File: examples/document_splitter_schema.py (hunk 1 above)
Comment (dswah) [reply examples/document_splitter_schema.py:22]: i think i recall that we had decided against the de/serialization when using the `serve` approach since it was superfluous: the pyfunc didnt need to receive a serialized payload (and the deserialize it) if the wholeapp runs on the same endpoint.

i think this de/serialization is only necessary when using a sidecar?

Comment (squat) [reply examples/document_splitter_schema.py:22]: Yes I spent way too much time on Friday only to re-learn that in the evening. Let me push my latest work, which actually raises some more questions.

Tldr, we indeed luckily don't need to de/serialize the request but in some cases we do have to re-shape it. I created CI that uses the mlflow example model [0] and I can only make it work if the schema transforms the request into an actual pandas dataframe object; a dict that has the same-ish shape as a dataframe isn't good enough. It seems like some models are very strict about what they accept.

[0] https://github.com/mlflow/mlflow-example

**Comment by squat** [comment]
>what to do with the request.transform and the json.dumps ? I recall that we had opted to remove that bit since we could achieve the desired behavior without encoding/decoding (which is less confusing too)

You're right, I think that this was simply an out of date example schema. I cleaned up that example schema :+1: 

> how do you update the schema.json files after making a change to one of the Schema.py files? is this automated? If not, what is the process? (and is the process documented in the readme?

The schema.json files are all generated automatically using the `meowlflow openapi` command. I added a check in the CI to ensure that the schema.json in the examples directory is kept up to date and added an e2e test to verify that the schema generation works as expected.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -17,12 +17,22 @@ jobs:
...
+        python-version: '3.9'
+        cache: poetry
+    - name: Install dependencies
+      run: poetry install
     - name: Docs
       run: |
+        poetry run meowlflow openapi --schema-path=examples/document_splitter_schema.py > examples/document_splitter_schema.json
         make README.md -B
-        DIFF="$(git status --porcelain 2>/dev/null )"
-        if [ "$(echo -n "$DIFF" | wc -l)" -ne 0 ]; then
```
Comment (squat) [inline .github/workflows/ci.yml:24]: `if [ "$(echo -n "$DIFF" | wc -l)" -ne 0 ]` seems to have returned false if the `$DIFF` was 0 or 1 lines long, meaning a change to a single file wouldn't prompt an error. More simply checking if the variable is empty using `if [ -z "$DIFF" ]` is more effective and correctly detects a single file change.

Comment (dswah) [reply .github/workflows/ci.yml:24]: 👌

File: e2e/mlflow_example_schema.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    "pH": "pH",
+    "residual_sugar": "residual sugar",
+    "sulphates": "sulphates",
+    "total_sulfur_dioxide": "total sulfur dioxide",
+    "volatile_acidity": "volatile acidity",
+}
+
+
+class Request(BaseRequest):
+    __root__: List[Properties]
```
Comment (dswah) [inline e2e/mlflow_example_schema.py:51]: wow cool example man!

File: examples/document_splitter_schema.py (hunk 1 above)
Comment (dswah) [reply examples/document_splitter_schema.py:21]: wow. very tricky!

File: meowlflow/sidecar.py [hunk 6]
```diff
@@ -2,12 +2,12 @@
 import time
 import logging
 import importlib.util
-from typing import TYPE_CHECKING
+from typing import TYPE_CHECKING, Any, Awaitable, Callable
```
Comment (dswah) [inline meowlflow/sidecar.py:5]: dope typing

File: meowlflow/sidecar.py [hunk 7]
```diff
@@ -114,17 +116,20 @@ def sidecar(
...
-        app,
+        app,  # nomypy, https://github.com/tiangolo/fastapi/issues/3927
         host=host,
         port=port,
         log_level="debug",
     )
 
 
-def get_infer(upstream):
+Infer = Callable[[Any], Awaitable[Any]]
```
Comment (dswah) [inline meowlflow/sidecar.py:126]: wow cool!

**Review by dswah — APPROVED** [review]
looks excellent! 
great work!

my only suggestion would be to add an entry in the readme about running the `meowlflow openapi` command, eg when and why to use it.

File: meowlflow/sidecar.py (hunk 7 above)
Comment (squat) [reply meowlflow/sidecar.py:126]: ❤️

File: meowlflow/sidecar.py (hunk 6 above)
Comment (squat) [reply meowlflow/sidecar.py:5]: ❤️

File: e2e/mlflow_example_schema.py (hunk 5 above)
Comment (dswah) [reply e2e/mlflow_example_schema.py:51]: right it would be a great example of a very standard dataset

**Comment by squat** [comment]
> my only suggestion would be to add an entry in the readme about running the meowlflow openapi command, eg when and why to use it.

yes that would be nice additional documentation. since the `openapi` command isn't new, I'll merge this for now and make another PR later to introduce the documentation. we should actually the `meowlflow build` and `meowlflow promote` commands too, now that I think about it.

File: pyproject.toml [hunk 8]
```diff
@@ -22,6 +22,9 @@ flake8 = "^4.0.1"
 black = "^22.1.0"
 pytest = "^6.2.5"
 pytest-cov = "^3.0.0"
+mypy = "^0.971"
+sklearn = "^0.0"
```
Comment (squat) [inline pyproject.toml:26]: sklearn and pandar are needed for the e2e tests to run the example model, thus adding them as development dependencies.

Comment (dswah) [reply pyproject.toml:26]: gebius
